### PR TITLE
Workaround for pytorch/pytorch#138575 distributed checkpoint loading bug

### PR DESCRIPTION
**Pull request by arjvik**
Upstream PyTorch has a bug (pytorch/pytorch#138575) with `dcp.load()` that causes the objects pointed to by `state_dict` to diverge from the objects in use by the train loop. Specifically, torch is supposed to update in-place the Stateful elements in `state_dict`, however it also replaces the references in `state_dict` with the newly loaded object, resulting in the CheckpointManager instance having a stale trainstate/optimizer/etc.

This causes an issue when attempting to load a model from a checkpoint and subsequently save the checkpoint -- the loaded trainstate/optimizer/etc are saved instead of the current values. As a result, if a training run is pre-empted and resumed twice, the results are subtly wrong.

Until pytorch/pytorch#138575 is merged, this should work around the issue.

From @karan-dalal and I, thanks for all the work on TorchTitan! We're glad to be able to contribute back upstream in return!

### CODE REVIEW

**Comment by fegin** [comment]
Thanks for the fix. Please fix the linter. We should land this first as it can take a while to land PyTorch fix.

File: torchtitan/checkpoint.py [hunk 1]
```diff
@@ -469,6 +469,13 @@ def load(self, step: int = -1) -> bool:
 
         # We won't have optimizer states to load, if we are loading a seed checkpoint
         states = {"model": self.states["model"]} if step == 0 else self.states
+        # PyTorch bug: (pytorch/pytorch#138575)
+        # dcp.load() replaces the values of stateful elements in `states` with new objects
+        # from loading the checkpointm, in addition to updating the states of the original
```
Comment (awgu) [inline torchtitan/checkpoint.py]: nit (typo?):
```suggestion
        # from loading the checkpoint, in addition to updating the states of the original
```

**Comment by arjvik** [comment]
@fegin fixed the linter warning in checkpoint.py

`ufmt` also wants changes to `torchtitan/datasets/tokenizer/tokenizer.py` and `torchtitan/models/llama_multimodal/model.py` which I didn't touch - left those unchanged for now, do you want me to add those to this PR?

**Comment by fegin** [comment]
We can land the PR once all tests are green.

**Comment by arjvik** [comment]
@fegin looks like it's all good!

**Review by fegin — APPROVED** [review]

**Comment by arjvik** [comment]
:tada:

**Pull request merged**